### PR TITLE
REFACTOR-#3994: remove code duplication for `get_indices` function

### LINKED ISSUE #3994: REFACTOR: remove code duplication for `get_indices` function

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

For both `Ray` and `Dask`, the base version of `get_indices` (in `PandasDataframePartitionManager`) is enough (if we create a new function that gets objects wrapped by "partitions" in parallel).

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/developer/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3994 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by dchigarev** [comment]
Isn't the base version of `get_indices` is sequential since it does `.get()` in a for-loop, which blocks functions submiting for other partitions?
https://github.com/modin-project/modin/blob/ed2a7a4adf973660eb15b9ea674cd5fb802f2424/modin/core/dataframe/pandas/partitioning/partition_manager.py#L795
Overwriden `get indices` first submits all functions and then gather the info from all futures at once with the engine-specific method.

**Comment by anmyachev** [comment]
@dchigarev ready for review

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 1]
```diff
@@ -777,6 +777,28 @@ def from_arrow(cls, at, return_dims=False):
...
+
+        Returns
+        -------
+        list
+            The objects wrapped by `partitions`.
+
+        Notes
+        -----
+        This method can be more efficient implemented for engines that support
+        getting objects in parallel.
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:798]:
```suggestion
        This method should be implemented in a more efficient way for engines that support
        getting objects in parallel.
```

**Review by vnlitvinov — APPROVED** [review]
LGTM!

And thanks @dchigarev for catching this.

**Review by dchigarev — APPROVED** [review]
LGTM

**Comment by anmyachev** [comment]
> Don't we try to hand over a lot of responsibilities to `PandasDataframePartitionManager`, whereas adding a new utilities that do not look beneficial? Not sure that new introduced method looks good to me. Given the phrase in the PR description
> 
> > For both Ray and Dask, the base version of get_indices (in PandasDataframePartitionManager) is enough.
> 
> if the get_indices implementation looked the same for all the engines (as for broadcast_apply in #3996 ), I would be okay with the changes.

Taking multiple features in one call to the engine is a fairly popular functionality, it can be used in other places, for example [here](https://github.com/modin-project/modin/pull/3781/files#diff-df691150497994b67a2083a74c037219dde5e38f417833cc049e57df9268fb6eL59). So we should create this function definitely. 

Regarding the place of definition, I did not find a better place than `PandasDataframePartitionManager`class, given the reference to the internal `oid/futures` field of the partition.

**Comment by anmyachev** [comment]
@dchigarev CI is green

**Review by dchigarev — APPROVED** [review]
LGTM, leaving the honor to merge for @YarShev as he had some questions

**Comment by YarShev** [comment]
> Taking multiple features in one call to the engine is a fairly popular functionality, it can be used in other places, for example [here](https://github.com/modin-project/modin/pull/3781/files#diff-df691150497994b67a2083a74c037219dde5e38f417833cc049e57df9268fb6eL59). So we should create this function definitely.

Given that, I am getting convinced we should have such a function.

File: modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition_manager.py [hunk 2]
```diff
@@ -35,45 +34,20 @@ class PandasOnDaskDataframePartitionManager(PandasDataframePartitionManager):
...
 
         Parameters
         ----------
-        axis : {0, 1}
-            Axis to extract the labels over.
         partitions : np.ndarray
-            The array of partitions from which need to extract the labels.
-        index_func : callable
-            The function to be used to extract the indices.
+            NumPy array with ``PandasDataframePartition``-s.
```
Comment (YarShev) [inline modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition_manager.py:44]:
```suggestion
            NumPy array with ``PandasOnDaskDataframePartition``-s.
```

File: modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition_manager.py [hunk 3]
```diff
@@ -35,45 +34,20 @@ class PandasOnDaskDataframePartitionManager(PandasDataframePartitionManager):
...
-        else:
-            new_idx = (
-                [idx.apply(func).future for idx in partitions[0]]
-                if len(partitions)
-                else []
-            )
-        new_idx = client.gather(new_idx)
-        return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
+        futures = [partition.future for partition in partitions]
+        return client.gather(futures)
```
Comment (YarShev) [inline modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition_manager.py:53]:
```suggestion
        return client.gather([partition.future for partition in partitions])
```

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py [hunk 4]
```diff
@@ -95,47 +94,22 @@ class PandasOnRayDataframePartitionManager(GenericRayDataframePartitionManager):
...
         """
-        Get the internal indices stored in the partitions.
+        Get the objects wrapped by `partitions` in parallel.
 
         Parameters
         ----------
-        axis : {0, 1}
-            Axis to extract the labels over.
         partitions : np.ndarray
             NumPy array with ``PandasDataframePartition``-s.
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py:104]:
```suggestion
            NumPy array with ``PandasOnRayDataframePartition``-s.
```

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py [hunk 5]
```diff
@@ -95,47 +94,22 @@ class PandasOnRayDataframePartitionManager(GenericRayDataframePartitionManager):
...
-        else:
-            new_idx = (
-                [idx.apply(func).oid for idx in partitions[0]]
-                if len(partitions)
-                else []
-            )
-        new_idx = ray.get(new_idx)
-        return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
+        futures = [partition.oid for partition in partitions]
+        return ray.get(futures)
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py:112]:
```suggestion
        return ray.get([partition.oid for partition in partitions])
```

**Comment by YarShev** [comment]
> > Taking multiple features in one call to the engine is a fairly popular functionality, it can be used in other places, for example [here](https://github.com/modin-project/modin/pull/3781/files#diff-df691150497994b67a2083a74c037219dde5e38f417833cc049e57df9268fb6eL59). So we should create this function definitely.
> 
> Given that, I am getting convinced we should have such a function.

Please fix those places with a new method as part of a separate issue when these changes are merged.

**Comment by anmyachev** [comment]
> > > Taking multiple features in one call to the engine is a fairly popular functionality, it can be used in other places, for example [here](https://github.com/modin-project/modin/pull/3781/files#diff-df691150497994b67a2083a74c037219dde5e38f417833cc049e57df9268fb6eL59). So we should create this function definitely.
> > 
> > 
> > Given that, I am getting convinced we should have such a function.
> 
> Please fix those places with a new method as part of a separate issue when these changes are merged.

ok

**Review by YarShev — DISMISSED** [review]

**Review by YarShev — APPROVED** [review]

**Pull request merged**